pkg/dao/crd/dao.go

```diff
@@ -311,78 +315,163 @@ func (d *Dao) SetState(instanceID string, state apb.JobState) (string, error) {
 
 // GetState - Retrieve a job state from the kvp API for an ID and Token.
 func (d *Dao) GetState(id string, token string) (apb.JobState, error) {
-	js, err := d.client.JobStates(d.namespace).Get(token, metav1.GetOptions{})
-	if err != nil {
-		log.Errorf("unable to get state for token: %v", err)
-		return apb.JobState{}, err
-	}
-	j, err := crd.ConvertJobStateToAPB(js.Spec, js.GetName())
-	if err != nil {
-		return apb.JobState{}, err
-	}
-	return *j, nil
+	// get the binding based on instance ID //update the job based on the token.
+	var job v1.Job
+	bi, err := d.client.BundleBindings(d.namespace).Get(id, metav1.GetOptions{})
+	if err != nil && !d.IsNotFoundError(err) {
+		log.Errorf("Unable to update the job state: %v - %v", token, err)
+		return apb.JobState{}, fmt.Errorf("unable to find job state %v", token)
+	} else if d.IsNotFoundError(err) {
+		si, err := d.client.BundleInstances(d.namespace).Get(id, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("Unable to update the job state: %v - %v", token, err)
+			return apb.JobState{}, err
+		}
+		if si.Status.Jobs == nil {
+			log.Errorf("Unable to update the job state: %v - %v", token, err)
+			return apb.JobState{}, err
+		}
+		j, ok := si.Status.Jobs[token]
+		if !ok {
+			log.Errorf("Unable to update the job state: %v - %v", token, err)
+			return apb.JobState{}, fmt.Errorf("unable to find job state %v", token)
+		}
+		job = j
+	} else {
+		if bi.Status.Jobs == nil {
+			log.Errorf("Unable to update the job state: %v - %v", token, err)
+			return apb.JobState{}, err
+		}
+		j, ok := bi.Status.Jobs[token]
+		if !ok {
+			log.Errorf("Unable to update the job state: %v - %v", token, err)
+			return apb.JobState{}, fmt.Errorf("unable to find job state %v", token)
+		}
+		job = j
+	}
+	return apb.JobState{
+		Description: job.Description,
+		Method:      crd.ConvertJobMethodToAPB(job.Method),
+		Podname:     job.Podname,
+		Token:       token,
+		State:       crd.ConvertStateToAPB(job.State),
+		Error:       job.Error,
+	}, nil
 }
 
 // GetStateByKey - Retrieve a job state from the kvp API for a job key
 func (d *Dao) GetStateByKey(key string) (apb.JobState, error) {
-	return d.GetState("", key)
+	bi, err := d.client.BundleBindings(d.namespace).Get(key, metav1.GetOptions{})
```

Comment:
Why does `GetStateByKey` only care about bindings? The usage in the broker only cares about that case?

Reply:
Yes that is the case, the only thing that has a key is the binding as well. apb.BindIntance